pandas/core/groupby/ops.py

```diff
@@ -1260,10 +1260,17 @@ def __iter__(self) -> Iterator:
         starts, ends = lib.generate_slices(self._slabels, self.ngroups)
 
         for start, end in zip(starts, ends):
-            yield self._chop(sdata, slice(start, end))
+            sliced = self._chop(sdata, slice(start, end))
+            if self._sorted_data is self.data:
+                yield sliced.copy()
+            else:
+                yield sliced
 
     @cache_readonly
     def _sorted_data(self) -> NDFrameT:
+        if lib.is_range_indexer(self._sort_idx, len(self.data)):
+            # Check if DF already in correct order
+            return self.data  # Need to take a copy later!
```

Comment:
can you flesh out this comment a little bit.  This is the sort of thing where I would preface a comment with "NB:"

Reply:
The reason to not make a copy here (once) is for memory reasons (to not have a full copy at once in memory)? 
Because in the end the full data will be copied as well when copying each slice? Just that the same amount of data gets copied in multiple copy() calls, and thus introducing more overhead.

Reply:
Yeah, that's basically the idea behind it.

For CoW, this is already optimized, since take will return a shallow copy.
I'm trying to see if this is viable for non CoW because I'm trying to apply this optimization to unsorted arrays as well(idea is to only take a chunk of the unsorted array at a time instead of the whole thingwith the argsort indices).

So far, I've gotten the copy() call overhead down to ~1.5 from 2.3. 
Most of the overhead that's left is from pandas itself (e.g. creating a view of the Index, BlockManager stuff), and the actual copy of the group values is relatively cheap.